### PR TITLE
localstore reserve logic

### PR DESCRIPTION
**Pull request by zelig**
the pr also fixes an index leak which happened (ModeGetRequest) when an as yet unsynced chunk was requested. It was put to GC incorrectly but moreover this entry was not removed when the chunk was synced (setSync was called) 

#1346 

AP https://hackmd.io/9ejVZEvSQg69zXG91YsFEQ

### CODE REVIEW

**Comment by acud** [comment]
@zelig adding a ton of renaming to fields and methods unnecessarily increases the mental taxation of reviewing an already large changeset. can you please revert all unrelated changes from the scope of this PR?

File: pkg/localstore/localstore.go [hunk 1]
```diff
@@ -276,9 +277,9 @@ func New(path string, baseKey []byte, o *Options, logger logging.Logger) (db *DB
 		return nil, err
 	}
 	// pull index allows history and live syncing per po bin
-	db.pullIndex, err = db.shed.NewIndex("PO|BinID->Hash|Tag", shed.IndexFuncs{
+	db.pullIndex, err = db.shed.NewIndex("PO|BinID->Hash|BatchID", shed.IndexFuncs{
```
Comment (acud) [inline pkg/localstore/localstore.go:280]: not sure, why is this needed if the information is already on the retrievalDataIndex?

File: pkg/localstore/mode_put.go [hunk 2]
```diff
@@ -85,13 +85,6 @@ func (db *DB) put(mode storage.ModePut, chs ...swarm.Chunk) (exist []bool, err e
 			}
 			exist[i] = exists
 			gcSizeChange += c
-
-			if mode == storage.ModePutRequestPin {
```
Comment (acud) [inline pkg/localstore/mode_put.go:89]: I'm reinstating this part since its needed, but I'm not sure how this will behave with the `gcSizeChange` line at 87

File: pkg/localstore/postage.go [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+		return err
+	}
+	item.Radius = radius
+	if err = db.postage.counts.PutInBatch(batch, item); err != nil {
+		return err
+	}
+	return db.shed.WriteBatch(batch)
+}
+
+func (p *postageBatches) withinRadius(item shed.Item) bool {
```
Comment (zelig) [inline pkg/localstore/postage.go:200]: oops this needs to be:

    return p.po(item.Address()) >= item.Radius

File: pkg/node/node.go [hunk 4]
```diff
@@ -258,7 +258,6 @@ func NewBee(addr string, swarmAddress swarm.Address, publicKey ecdsa.PublicKey,
 		return nil, fmt.Errorf("localstore: %w", err)
 	}
 	b.localstoreCloser = storer
```
Comment (zelig) [inline pkg/node/node.go:260]: there needs to be a change here. setting up the batchstore with localstore.Unreserve hood

File: pkg/localstore/mode_put.go [hunk 5]
```diff
@@ -259,21 +262,31 @@ func (db *DB) putSync(batch *leveldb.Batch, binIDs map[uint8]uint64, item shed.I
...
+// (the retrieval access index and the gc index)
+func (db *DB) preserveOrCache(batch *leveldb.Batch, item shed.Item, forcePin bool) (gcSizeChange int64, err error) {
+	// item needs to be populated with Radius
+	item2, err := db.postageRadiusIndex.Get(item)
+	if err != nil {
+		return 0, fmt.Errorf("postage radius index: %w", err)
+	}
+	item.Radius = item2.Radius
+
+	if withinRadiusFn(db, item) || forcePin {
```
Comment (zelig) [inline pkg/localstore/mode_put.go:287]: when this is from a simple setsync call we also should not set the access timestamp to now.
only if it is putRequest.

but surely not ike this, instead:
*make access update consistent irrespective of reserve/cache status*. (as reserve can be evicted after a second and then will be deleted even before their earlier cached (and earlier accessed) sister chunks.

File: pkg/localstore/reserve.go [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+		}
+		batch = new(leveldb.Batch)
+	)
+
+	i, err := db.postageRadiusIndex.Get(item)
+	if err != nil {
+		if !errors.Is(err, leveldb.ErrNotFound) {
+			return err
+		}
+		item.Radius = radius
```
Comment (zelig) [inline pkg/localstore/reserve.go:34]: the batch needs to save the new radius even if the batch IS found in the postageRadiusIndex.

File: pkg/localstore/reserve.go [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+			return false, fmt.Errorf("unpin: %w", err)
+		}
+
+		// if the batch is unreserved we should remove the chunk
+		// from the pull index
+		item2, err := db.retrievalDataIndex.Get(item)
+		if err != nil {
+			return false, err
+		}
+		err = db.pullIndex.DeleteInBatch(batch, item2)
```
Comment (zelig) [inline pkg/localstore/reserve.go:54]: this is not consistent. The pullindex must include chunk batches otherwise nothing will be in the pullindex from any bin outside the neighbourhood radius

File: pkg/localstore/mode_put.go [hunk 8]
```diff
@@ -307,11 +310,17 @@ func (db *DB) setGC(batch *leveldb.Batch, item shed.Item) (gcSizeChange int64, e
 		return 0, err
 	}
 	if !ok {
-		err = db.gcIndex.PutInBatch(batch, item)
-		if err != nil {
+		exists, err := db.gcIndex.Has(item)
```
Comment (zelig) [inline pkg/localstore/mode_put.go:313]:
```go
if ok {
    return gcSizeChange, nil
}
exists, err := db.gcIndex.Has(item)
```

File: pkg/localstore/mode_put.go [hunk 9]
```diff
@@ -307,11 +310,17 @@ func (db *DB) setGC(batch *leveldb.Batch, item shed.Item) (gcSizeChange int64, e
...
+		if err != nil && !errors.Is(err, leveldb.ErrNotFound) {
 			return 0, err
 		}
-		gcSizeChange++
+		if !exists {
```
Comment (zelig) [inline pkg/localstore/mode_put.go:317]:
```
if exists {
    return 0, nil
}
```

File: pkg/localstore/mode_set.go [hunk 10]
```diff
@@ -179,65 +175,51 @@ func (db *DB) setSync(batch *leveldb.Batch, addr swarm.Address, mode storage.Mod
...
-	}
-
-	// Add in gcIndex only if this chunk is not pinned
-	ok, err := db.pinIndex.Has(item)
-	if err != nil {
-		return 0, err
-	}
-	if !ok {
-		err = db.gcIndex.PutInBatch(batch, item)
+		item.AccessTimestamp = now()
```
Comment (zelig) [inline pkg/localstore/mode_set.go:183]: TODO: here please take note of this. This should leave the access timestamp alone.

Comment (zelig) [reply pkg/localstore/mode_set.go:183]: and the following line also dropped

File: pkg/localstore/reserve_test.go [hunk 11]
```diff
@@ -0,0 +1,278 @@
...
+		err = db.Set(context.Background(), storage.ModeSetSync, ch.Address())
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	var gcChs []swarm.Chunk
+	for i := 0; i < 100; i++ {
+		gcch := generateTestRandomChunkAt(swarm.NewAddress(db.baseKey), 2).WithBatch(2, 3)
+		err := db.UnreserveBatch(gcch.Stamp().BatchID(), 2)
```
Comment (zelig) [inline pkg/localstore/reserve_test.go:212]: maybe a better test would be to add a few chunks of at PO 4 and show they are intact.
otherwise this is a very good test`

File: pkg/localstore/mode_get.go [hunk 12]
```diff
@@ -151,26 +151,28 @@ func (db *DB) updateGC(item shed.Item) (err error) {
...
-	ok, err := db.pinIndex.Has(item)
-	if err != nil {
-		return err
-	}
-	if !ok {
+	updatedTimestamp := false
+	_, err = db.gcIndex.Get(item)
+	if err == nil {
+		updatedTimestamp = true
+		item.AccessTimestamp = now()
```
[the remark this answers is not in the document]
Comment (acud) [reply pkg/localstore/mode_get.go:159]: look 9 lines above?

File: pkg/localstore/mode_put.go [hunk 13]
```diff
@@ -259,46 +271,37 @@ func (db *DB) putSync(batch *leveldb.Batch, binIDs map[uint8]uint64, item shed.I
 	if err != nil {
 		return false, 0, err
 	}
-	gcSizeChange, err = db.setGC(batch, item)
+	err = db.postageChunksIndex.PutInBatch(batch, item)
+	if err != nil {
+		return false, 0, err
+	}
+
+	item.AccessTimestamp = now()
```
[the remark this answers is not in the document]
Comment (acud) [reply pkg/localstore/mode_put.go:279]: ?

File: pkg/localstore/mode_set.go (hunk 10 above)
Comment (acud) [reply pkg/localstore/mode_set.go:183]: ....don't understand.... i thought you said setSync should update access timestamp

Comment (acud) [reply pkg/localstore/mode_set.go:183]: let's please have a writeup or at least a truth table about what happens to which index under which circumstances. i find documenting this in the code not the best thing to do and it just pollutes the code with hypothetical reminders. let's start with defining, then make changes to the codebase and in the meanwhile try to keep it clean

Comment (zelig) [reply pkg/localstore/mode_set.go:183]: here it is quite simple, you update accesstimestamp if and only if the chunk iis requested

**Review by zelig — COMMENTED** [review]
lgtm

File: pkg/localstore/mode_set.go (hunk 10 above)
Comment (acud) [reply pkg/localstore/mode_set.go:183]: truth table?

**Review by acud — APPROVED** [review]

**Comment by acud** [comment]
famous last words?
